### PR TITLE
firefox browser

### LINKED ISSUE #236: Implement protocol generator for Firefox

### PR DESCRIPTION
**Pull request by Meir017**
closes #236 

- [x] protocol generation
- [x] firefox connection
- [x] firefox session
- [x] first passing test using ws transport

### CODE REVIEW

**Comment by Meir017** [comment]
@kblok protocol generation is working now. firefox is weired :(

**Comment by kblok** [comment]
> @kblok protocol generation is working now. firefox is weired :(

haha what happened?

Did you keep track of my changes in the generator? There are more coming on #252

File: src/PlaywrightSharp/Protocol/IProtocolEvent.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+{
+    /// <summary>
+    /// Base type for protocol events.
+    /// </summary>
+    internal interface IProtocolEvent
+    {
+        /// <summary>
+        /// The event name raised by the devtools protocol.
+        /// </summary>
+        [JsonIgnore]
```
[the remark this answers is not in the document]
Comment (Meir017) [reply src/PlaywrightSharp/Protocol/IProtocolEvent.cs]: oops. I just copied it from the chromium...

File: src/PlaywrightSharp.ProtocolTypesGenerator/Firefox/FirefoxProtocolTypesGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                    _ => throw new ArgumentOutOfRangeException(nameof(obj), type, $"[{domain}.{name}] invalid object type")
+                };
+            }
+            catch (Exception e)
+            {
+                throw new Exception("ERROR", e);
+            }
+        }
+
+        // INodeServices is deprecated...
```
[the remark this answers is not in the document]
Comment (Meir017) [reply src/PlaywrightSharp.ProtocolTypesGenerator/Firefox/FirefoxProtocolTypesGenerator.cs]: there is but I was too lazy to find it... ;)

**Comment by kblok** [comment]
@Meir017 can you take a look at the failing tests?

File: src/PlaywrightSharp.Abstractions/BrowserArgOptions.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ public class BrowserArgOptions
         /// <summary>
         /// Whether to run browser in headless mode. Defaults to true unless the devtools option is true.
         /// </summary>
-        public bool Headless { get; set; }
+        public bool? Headless { get; set; }
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/BrowserArgOptions.cs]: I would leave this as `bool` and I would default this to true.

File: src/PlaywrightSharp.Abstractions/BrowserArgOptions.cs [hunk 4]
```diff
@@ -25,6 +25,6 @@ public class BrowserArgOptions
         /// <summary>
         /// Whether to auto-open DevTools panel for each tab. If this option is true, the headless option will be set false.
         /// </summary>
-        public bool Devtools { get; set; }
+        public bool? Devtools { get; set; }
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/BrowserArgOptions.cs]: We could mimic pptr-sharp here.

File: src/PlaywrightSharp.Firefox/FirefoxConnection.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                await RawSendAsync(new ConnectionRequest
+                {
+                    Id = id,
+                    Method = request.Command,
+                    Params = request,
+                }).ConfigureAwait(false);
+            }
+
+            // We need to silence exceptions on async void events.
+#pragma warning disable CA1031 // Do not catch general exception types.
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxConnection.cs]: I removed this rue, we can get rid of this.

File: src/PlaywrightSharp.Firefox/FirefoxProcessManager.cs [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+using System.Diagnostics;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using PlaywrightSharp.Helpers;
+
+namespace PlaywrightSharp.Firefox
+{
+    internal class FirefoxProcessManager : IProcessManager
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxProcessManager.cs]: @Meir017 do you think we can reuse this `ProcessManager` among browsers?

File: src/PlaywrightSharp.ProtocolTypesGenerator/Firefox/FirefoxProtocolTypesGenerator.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meir017) [reply src/PlaywrightSharp.ProtocolTypesGenerator/Firefox/FirefoxProtocolTypesGenerator.cs]: they stopped supporting it but it still works https://github.com/dotnet/aspnetcore/issues/12890

File: src/PlaywrightSharp.Firefox/FirefoxProcessManager.cs (hunk 6 above)
Comment (Meir017) [reply src/PlaywrightSharp.Firefox/FirefoxProcessManager.cs]: I'm sure we can. we will need to add some abstraction for that. firefox gets the ws url from the output data and not the error data but it's makes sense to use the same logic

Comment (kblok) [reply src/PlaywrightSharp.Firefox/FirefoxProcessManager.cs]: I will create an issue for that, we will have a better picture when we implement Webkit.

File: src/PlaywrightSharp.Firefox/FirefoxConnection.cs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+                    System.Diagnostics.Debug.WriteLine($"{ex}: Failed to deserialize response {response}");
+                    return;
+                }
+
+                // _logger.LogTrace("◀ Receive {Message}", response);
+                ProcessIncomingMessage(obj);
+            }
+
+            // We need to silence exceptions on async void events.
+#pragma warning disable CA1031 // Do not catch general exception types.
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxConnection.cs]: We can remove this

File: src/PlaywrightSharp.Firefox/FirefoxTarget.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+        /// <inheritdoc cref="ITarget.GetPageAsync"/>
+        public Task<IPage> GetPageAsync() => Task.FromResult<IPage>(null);
+
+        public Task<IWorker> GetWorkerAsync()
+        {
+            throw new NotImplementedException();
+        }
+
+        internal void DidClose()
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxTarget.cs]: Arrow

File: src/PlaywrightSharp.Firefox/FirefoxSession.cs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                {
+                    Id = id,
+                    Method = request.Command,
+                    Params = request,
+                    SessionId = _sessionId,
+                }).ConfigureAwait(false);
+            }
+
+            // We need to silence exceptions on async void events.
+#pragma warning disable CA1031 // Do not catch general exception types.
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxSession.cs]: We can remove this

File: src/PlaywrightSharp.Firefox/FirefoxSession.cs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                }
+                else
+                {
+                    var result = FirefoxProtocolTypes.ParseResponse(callback.Method, obj.Result.Value.GetRawText());
+                    callback.TaskWrapper.TrySetResult(result);
+                }
+            }
+        }
+
+        internal void OnMessageReceived(object sender, IFirefoxEvent e)
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxSession.cs]: Arrow

File: src/PlaywrightSharp.Firefox/FirefoxTarget.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        public TargetType Type { get; }
+
+        public IBrowserContext BrowserContext => throw new NotImplementedException();
+
+        ITarget ITarget.Opener => Opener;
+
+        internal bool IsInitialized { get; set; }
+
+        internal FirefoxTarget Opener => _openerId != null ?
+            _browser.TargetsMap[_openerId] : null;
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxTarget.cs]: This can fit in one line

File: src/PlaywrightSharp/Protocol/ProtocolTypes.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        where TProtocolEvent : IProtocolEvent
+    {
+        private static readonly IDictionary<string, Type> ResponsesMapper = new Dictionary<string, Type>();
+        private static readonly IDictionary<string, Type> EventsMapper = new Dictionary<string, Type>();
+
+        static ProtocolTypes()
+        {
+            var chromiumResponse = typeof(TProtocolResponse);
+            var chromiumRequest = typeof(TProtocolRequest);
+            var chromiumEvent = typeof(TProtocolEvent);
```
Comment (kblok) [inline src/PlaywrightSharp/Protocol/ProtocolTypes.cs]: Can we remove `chromium` naming here?

**Comment by Meir017** [comment]
@kblok tests aren't failing now

**Pull request merged**